### PR TITLE
Avoid repeated trait bounds in derived impls

### LINKED ISSUE #25022: rustdoc shows duplicate bounds for derived implementations
See the derived implementations at the bottom of http://doc.rust-lang.org/1.0.0-beta.3/std/num/struct.Wrapping.html for example.

```rust
impl<T> Clone for Wrapping<T> where T: Clone + Clone
```

### PR DESCRIPTION
**Pull request by fhartwig**
Fixes #25022

This adapts the deriving mechanism to not repeat bounds for the same type parameter. To give an example: for the following code:

``` rust
#[derive(Clone)]
pub struct FlatMap<I, U: IntoIterator, F> {
    iter: I,
    f: F,
    frontiter: Option<U::IntoIter>,
    backiter: Option<U::IntoIter>,
}
```

the latest nightly generates the following impl signature:

``` rust
impl <I: ::std::clone::Clone,
      U: ::std::clone::Clone + IntoIterator,
      F: ::std::clone::Clone>
::std::clone::Clone for FlatMap<I, U, F> where
    I: ::std::clone::Clone,
    F: ::std::clone::Clone,
    U::IntoIter: ::std::clone::Clone,
    U::IntoIter: ::std::clone::Clone
```

With these changes, the signature changes to this:

``` rust
impl <I, U: IntoIterator, F> ::std::clone::Clone for FlatMap<I, U, F> where
    I: ::std::clone::Clone,
    F: ::std::clone::Clone,
    U::IntoIter: ::std::clone::Clone
```

(Nothing in the body of the impl changes)
Note that the second impl is more permissive, as it doesn't have a `Clone` bound on `U` at all. There was a compile-fail test that failed due to this. I don't understand why we would want the old behaviour (and nobody on IRC could tell me either), so please tell me if there is a good reason that I missed.

### CODE REVIEW

File: src/libsyntax/ext/deriving/generic/mod.rs [hunk 1]
```diff
@@ -548,11 +546,17 @@ impl<'a> TraitDef<'a> {
...
                 .collect();
+            let mut processed_types = HashSet::new();
 
             for field_ty in field_tys {
                 let tys = find_type_parameters(&*field_ty, &ty_param_names);
 
                 for ty in tys {
+                    // if we have already handled this type, skip it
+                    if let ast::TyPath(_, ref p) = ty.node {
+                        if processed_types.contains(&p.segments) { continue }
```
Comment (huonw) [inline src/libsyntax/ext/deriving/generic/mod.rs]: This can be `if !processed_types.insert(&p.segments) { continue }`.

**Comment by huonw** [comment]
The old behaviour is because it was a very simple rule (`#[derive]` puts the trait as a bound on each param) and is somewhat more consistent (at least, can be interpreted as more consistent): structs like `struct Foo<T>(Rc<T>);` or `struct Foo<T>(PhantomData<T>);` don't need `T: Clone`, but `#[derive(Clone)]` will add it anyway, even though they're "unused".

I don't really know where to go... maybe we can preserve the old behaviour while still removing the double trait bounds? (E.g. check if type parameters are in the `processed_types` set and add bounds if they're not?)

**Comment by huonw** [comment]
(I'd r+ in a heartbeat if the patch just deduplicates, I'm unsure about the other changes... I want it, but it'd be nice to do it "properly", so that the examples I mention above are also handled but it's unclear how we can do this best.)

**Comment by fhartwig** [comment]
Yeah, I can see the consistency argument. Fixing the bounds "properly" would be a much bigger change, and I'm not actually sure it can be done with `derive` operating at a purely syntactic level.
I'll change this to keep the old behaviour.
I've also considered not touching the deriving mechanism at all, and just cleaning up redundant bounds in rustdoc. Avoiding redundant bounds in the first place seemed like the cleaner way to handle this, but now I'm not sure anymore. What do you think?

**Comment by huonw** [comment]
> I'm not actually sure it can be done with derive operating at a purely syntactic level.

It can sort-of be done by instead inserting `where` clauses for each field, e.g. the example you give would just have `where F: Clone, I: Clone, Option<U::IntoIter>: Clone`. However, this reveals implementation details in an undesirable way. Other than that, yes, it'd need to be more than syntactic.

> What do you think?

It's something that probably makes sense to handle in rustdoc too but this is a more general problem: it could be much better at converting source code into a "canonical" representation.

**Comment by fhartwig** [comment]
@huonw Ok, I have amended my changes to keep the bounds on all type parameters. The example signature now looks as follows:

```rust
impl <I: ::std::clone::Clone,
      U: ::std::clone::Clone + IntoIterator,
      F: ::std::clone::Clone>
    ::std::clone::Clone for FlatMap<I, U, F>
    where U::IntoIter: ::std::clone::Clone {
```

File: src/libsyntax/ext/deriving/generic/mod.rs [hunk 2]
```diff
@@ -549,10 +550,20 @@ impl<'a> TraitDef<'a> {
...
 
+            let mut processed_field_types = HashSet::new();
             for field_ty in field_tys {
                 let tys = find_type_parameters(&*field_ty, &ty_param_names);
 
                 for ty in tys {
+                    // if we have already handled this type, skip it
+                    if let ast::TyPath(_, ref p) = ty.node {
+                        if p.segments.len() == 1
+                            && ty_param_names.contains(&p.segments[0].identifier.name)
```
Comment (huonw) [inline src/libsyntax/ext/deriving/generic/mod.rs]: Maybe we could just add these to the `process_fields_types` `HashSet` to begin with; `processed_field_types.extend(ty_param_names.iter().cloned())`.

Comment (fhartwig) [inline src/libsyntax/ext/deriving/generic/mod.rs]: I don't think so, unless I'm misunderstanding your comment. `ty_param_names` are `ast::Name`s, while the elements of `processed_field_types` are `Vec<ast::PathSegment>`. It's not clear to me that we can somehow convert one into the other for comparison.

**Pull request merged**